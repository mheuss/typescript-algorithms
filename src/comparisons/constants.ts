export enum Comparisons {
  LessThan = -1,
  Equal,
  GreaterThan,
}

export type ComparisonFunction<T> = (a: T, b: T) => Comparisons;
