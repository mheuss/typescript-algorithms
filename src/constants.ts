export enum ErrorCodes {
  COMPARISON_FUNCTION_REQUIRED = 1,
  LIST_IS_BROKEN_INTERNALLY,
  OPERATION_BEYOND_BOUNDS,
  FAILED_TO_SET,
  COMPARISON_ERROR,
  INVALID_TREE_DATA_FILE,
}
