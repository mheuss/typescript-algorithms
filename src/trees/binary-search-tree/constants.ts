export enum Sides {
  Left,
  Right,
}
