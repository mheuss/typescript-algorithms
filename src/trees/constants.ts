export interface ITreeObject<T> {
  children: Array<ITreeObject<T>>;
  value: T;
}

export enum Traversal {
  Breadth_First,
  DepthFirstInOrder,
  DepthFirstPreOrder,
  DepthFirstPostOrder,
}
