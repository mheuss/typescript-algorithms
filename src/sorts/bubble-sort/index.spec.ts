import { ErrorCodes } from "../../constants";
import { BubbleSort } from "./index";

describe("Bubble Sort Tests", () => {
  it("Should be able to sort a list of numbers in order", () => {
    const arr = [5, 2, 8, 67, 2, 5, 7, 4, 1, 7, 4, 67, 8, 9, 3];
    const results = BubbleSort<number>(arr);
    expect(results.length).toEqual(arr.length);
    expect(results).toMatchSnapshot();
  });

  it("Should be able to sort strings in correct order", () => {
    const arr = [
      "Michael",
      "Constance",
      "Mikey",
      "Lauren",
      "not a first name",
      "note",
    ];
    const results = BubbleSort<string>(arr);
    expect(results.length).toEqual(arr.length);
    expect(results).toMatchSnapshot();
  });

  it("Should be able to support an array of objects using a custom sort", () => {
    interface INames {
      lastName: string;
      firstName: string;
    }

    const arr = [
      { lastName: "Heuss", firstName: "Mike" },
      { lastName: "Smith", firstName: "John" },
      { lastName: "Arnold", firstName: "Peggy" },
      { lastName: "Saben", firstName: "Matt" },
      { lastName: "LaCroix", firstName: "Mark" },
    ];
    const results = BubbleSort<INames>(arr, (a: INames, b: INames) => {
      return a.lastName.localeCompare(b.lastName);
    });
    expect(results.length).toEqual(arr.length);
    expect(results).toMatchSnapshot();
  });

  it("Should return an empty array if given an empty array", () => {
    const results = BubbleSort<number>([]);
    expect(results).toEqual([]);
  });

  it("Should throw if array is of type it cannot sort, and no sort function was passed in", () => {
    const arr = [
      { lastName: "Heuss", firstName: "Mike" },
      { lastName: "Smith", firstName: "John" },
      { lastName: "Arnold", firstName: "Peggy" },
      { lastName: "Saben", firstName: "Matt" },
      { lastName: "LaCroix", firstName: "Mark" },
    ];
    try {
      BubbleSort(arr);
      fail("Should have thrown");
    } catch (e) {
      // @ts-expect-error Not worrying about types in unit test
      expect(e.code).toEqual(ErrorCodes.COMPARISON_FUNCTION_REQUIRED);
    }
  });
});
