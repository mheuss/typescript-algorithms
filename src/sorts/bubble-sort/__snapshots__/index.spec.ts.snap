// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bubble Sort Tests Should be able to sort a list of numbers in order 1`] = `
[
  1,
  2,
  2,
  3,
  4,
  4,
  5,
  5,
  7,
  7,
  8,
  8,
  9,
  67,
  67,
]
`;

exports[`Bubble Sort Tests Should be able to sort strings in correct order 1`] = `
[
  "Constance",
  "Lauren",
  "Michael",
  "Mikey",
  "not a first name",
  "note",
]
`;

exports[`Bubble Sort Tests Should be able to support an array of objects using a custom sort 1`] = `
[
  {
    "firstName": "Peggy",
    "lastName": "Arnold",
  },
  {
    "firstName": "Mike",
    "lastName": "Heuss",
  },
  {
    "firstName": "Mark",
    "lastName": "LaCroix",
  },
  {
    "firstName": "Matt",
    "lastName": "Saben",
  },
  {
    "firstName": "John",
    "lastName": "Smith",
  },
]
`;
